The -EACCESS error code set at the top never survives to the bottom
of the function.

I'm not sure we need all the syslogging going on here.

There are some self-explanatory comments (not only here).

Index: linux-2.6-apparmor/security/apparmor/lsm.c
===================================================================
--- linux-2.6-apparmor.orig/security/apparmor/lsm.c
+++ linux-2.6-apparmor/security/apparmor/lsm.c
@@ -594,19 +594,15 @@ static int apparmor_setprocattr(struct t
 	const char *cmd_changehat = "changehat ",
 		   *cmd_setprofile = "setprofile ";
 
-	int error = -EACCES;	/* default to a perm denied */
+	int error;
 	char *cmd = (char *)value;
 
-	/* only support messages to current */
-	if (strcmp(name, "current") != 0) {
-		error = -EINVAL;
+	error = -EINVAL;
+	if (strcmp(name, "current") != 0)
 		goto out;
-	}
-
-	if (!size) {
-		error = -ERANGE;
+	error = -ERANGE;
+	if (!size)
 		goto out;
-	}
 
 	/* CHANGE HAT -- switch task into a subhat (subprofile) if defined */
 	if (size > strlen(cmd_changehat) &&
@@ -631,7 +627,6 @@ static int apparmor_setprocattr(struct t
 
 		error = aa_setprocattr_changehat(hatinfo, infosize);
 		if (error == 0)
-			/* success, set return to #bytes in orig request */
 			error = size;
 
 	/* SET NEW PROFILE */
